File: 2020/day1/solve.js
const assert = require("assert");
const utils = require("../utils");

const wantedNumber = 2020;

function solver(input) {
  for (let i = 0; i < input.length; i++) {
    for (let j = 0; j < input.length; j++) {
      for (let k = 0; k < input.length; k++) {
        if ((i != j) != k && input[i] + input[j] + input[k] === wantedNumber) {
          return input[i] * input[j] * input[k];
        }
      }
    }
  }
}

assert(solver([1721, 979, 366, 299, 675, 1456]) == 241861950);

async function main() {
  const input = await utils.readInput("day1/input.txt", utils.types.INT);
  console.log(solver(input));
}

main();

File: 2023/day9/solve.go
package main

import (
	"fmt"
	"log"
	"saneri/aoc/utils"
	"strings"
)

func extrapolate(history []int, backwards bool) int {
	if (len(history) == 1) {
		log.Fatal("history is empty")
	}
	allZeros := true
	diffs := []int{}
	for i := range(history[1:]) {
		diff := history[i+1] - history[i]
		diffs = append(diffs, diff)
		if diff != 0 {
			allZeros = false
		}
	}
	if !allZeros {
		diffToNext := extrapolate(diffs, backwards)
		if (backwards) {
			return history[0] - diffToNext
		}
		return history[len(history)-1] + diffToNext
	}
	return history[0]
}

func main() {
	data := utils.ReadInput("input.txt")
	aSum := 0
	bSum := 0 
	for _, line := range data {
		history := utils.ArrayStringToInt(strings.Split(line, " "))
		aSum += extrapolate(history, false)
		bSum += extrapolate(history, true)
	}
	fmt.Println("a:", aSum)
	fmt.Println("b:", bSum)
}